### PR TITLE
#6029 fix broken multiline textboxes

### LINKED ISSUE #6029: Commit Template Settings: Commit Template field one line only
**Do you want to request a *feature* or report a *bug*?**
bug
**What is the current behavior?**
image
Try pressing enter or shift enter in textbox to add lines.
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
- In commit window, open 
image
- Try to type multiple lines

**What is the expected behavior?**
Be able to type multiple lines.
**Environment you encounter the issue:**

- Git Extensions 3.00.00.4433
- Build fca7cf228b481ee8c1b779cf7b882ccdfbdcd1bc
- Git 2.20.1.windows.1
- Microsoft Windows NT 6.3.9600.0
- .NET Framework 4.7.3260.0
- DPI 96dpi (no scaling)

**Did this work in previous version of GitExtensions (which)?**
Not sure.

### PR DESCRIPTION
**Pull request by NikolayXHD**
Fixes #6029

## Proposed changes
Handle missing `Ctrl`+`Backspace` hotkey by emulating `Ctrl`+`Shift`+`Left`, `Backspace` key sequence.
This approach works fine for both single-line and multiline textbox.

There is a  visual downside, one can see how the word becomes selected and then removed one moment after.

## Test methodology

Manually tested by editing multiline & single-line inputs in commit template window.

## Test environment(s)

Windows 7 SP1

### CODE REVIEW

**Comment by drewnoakes** [comment]
Should this workaround only be used for multiline text boxes?

File: GitExtUtils/GitUI/ControlHotkeyExtensions.cs [hunk 1]
```diff
@@ -24,31 +27,29 @@ public static void EnableRemoveWordHotkey(this Control control)
...
+        {
+            if (e.KeyData == (Keys.Control | Keys.Back))
+            {
+                e.Handled = true;
+                e.SuppressKeyPress = true;
 
-                        break;
-                }
+                // emulate keyboard sequence: Ctrl + Shift + Left, Backspace
+                SendKeys.Send("^+{LEFT} {BACKSPACE}");
```
Comment (vbjay) [inline GitExtUtils/GitUI/ControlHotkeyExtensions.cs]: https://stackoverflow.com/a/17913548

**Comment by NikolayXHD** [comment]
> Should this workaround only be used for multiline text boxes?

I would not do so for several reasons.

1. Possible unknown side effects of the previous fix besides the one we are fixing here.
1a. Known side effect: Enter key selects all text in the text box. Not harmful, but not an expected behavior either.
1b. In contrast with the current approach the one I suggest is explicit. Anyone can see what it does (or at least what it is expected to do) just by reading the code.
2. Simplicity, potential bugs are discovered and fixed twice as fast on average, because all the text boxes are affected the same way, instead of half of them one way while the other half another way.

File: GitExtUtils/GitUI/ControlHotkeyExtensions.cs (hunk 1 above)
Comment (NikolayXHD) [reply GitExtUtils/GitUI/ControlHotkeyExtensions.cs]: Fixed flickering by calling `Control.BeginUpdateInternal()` / `Control.EndUpdateInternal()`.

**Comment by RussKie** [comment]
@vbjay @gerhardol have you had a chance to test?

**Comment by vbjay** [comment]
No. We need to get my pull request in so portable builds run.

**Comment by RussKie** [comment]
@NikolayXHD it looks like we've haven't fixed it completely (note to self: test better)
As it was pointed at https://github.com/dotnet/winforms/pull/260#issuecomment-463636594
> if you hold Ctrl and press Backspace multiple times, only the first one results in the shortcut working properly, and subsequent presses are treated as a plain Backspace.

What do you think of the WinForms PR - https://github.com/dotnet/winforms/pull/260?

/cc: @zvirja

**Comment by NikolayXHD** [comment]
I checked my hobbyy app which also has text input supporting `Ctrl`+`Backspace`. Turns out `RichTextBox` handles `Ctrl`+`Backspace` properly.

**Comment by RussKie** [comment]
We just need to get it working (if it is possible, that is) for `TextBox` controls...

**Comment by NikolayXHD** [comment]
SUDDENLY I tested holding down `Ctrl` and repeatedly pressing `Backspace` with `TextBox` in Gitextensions in commit template (name field) and a whole word was removed every time.

It was my home PC with Win7. Today I will also test Win10 at work.

**Comment by RussKie** [comment]
i tested it on 3.0.2 in W10 in "Create branch" dialog and it would only
delete a single word, turning into delete single char after working once.
The key here is "press and hold CTRL" + "press BKSP repeatedly"
I'll post an animation tomorrow

**Comment by NikolayXHD** [comment]
> The key here is "press and hold CTRL" + "press BKSP repeatedly"

That's right, I was checking exactly that.

I also tested Create Branch and still cannot reproduce the issue.

I aslo remembered I happen to have a virtual machine with Win10 at home so I was able to test Win10 too, and still cannot reproduce the issue.

I type several words (with or without trailing whitespace)
I press and hold `Ctrl`
I press and release `Backspace` several times
I release `Ctrl`

every time a word (with subsequent backspace) is removed.

**Comment by RussKie** [comment]
I tried it on two W10 (one native, another under Parallels) and the result is the same - hold <kbd>CTRL</kbd> and repeatedly press <kbd>BKSP</kbd> - only the first word is deleted, then it turns into a normal <kbd>BKSP</kbd> button...

ctrl delete

**Comment by NikolayXHD** [comment]
Alghough I am still not able to get second `Backspace` remove 1 char only, I discovered yet another issue. If I use right `Ctrl` instead of left, then instead of removing the word the input gets into infinite loop of removing words.

**Comment by RussKie** [comment]
Wow...
Any thoughts on [dotnet/winforms#260](https://github.com/dotnet/winforms/pull/260)?

**Comment by NikolayXHD** [comment]
it lacks flicker prevention

**Pull request merged**